### PR TITLE
cluescrolls: Add Music Clues

### LINKED ISSUE #8497: Highlight the song for music clues
**Is your feature request related to a problem? Please describe.**
Find the specific music could take sometime, with a highlight where to find it, like the emote clue's it would speed up clues

**Describe the solution you'd like**
Highlight the music that is required for the cluescroll.
Highlight the arrow (up or down) where to find the specific song in the list.

**Additional context**
Could look like this

music_hint

_Maybe this is also a solution to find fairy ring codes faster.._

### PR DESCRIPTION
**Pull request by Hydrox6**
Closes #8497

java_2019-04-11_21-44-48

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import net.runelite.client.ui.overlay.components.LineComponent;
+import net.runelite.client.ui.overlay.components.PanelComponent;
+import net.runelite.client.ui.overlay.components.TitleComponent;
+import static net.runelite.client.plugins.cluescrolls.ClueScrollOverlay.TITLED_CONTENT_COLOR;
+import static net.runelite.client.plugins.cluescrolls.ClueScrollWorldOverlay.IMAGE_Z_OFFSET;
+
+@Getter
+public class MusicClue extends ClueScroll implements NpcClueScroll
+{
+	private static WorldPoint LOCATION = new WorldPoint(2990, 3384, 0);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java]: `final` for these

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+	public static MusicClue forText(String text)
+	{
+		final Matcher m = SONG_PATTERN.matcher(text);
+		m.find();
+		final String song = m.group(1);
+		return new MusicClue(song);
+	}
+
+	public String[] getNpcs()
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java]: ?

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+	@Override
+	public void makeWorldOverlayHint(Graphics2D graphics, ClueScrollPlugin plugin)
+	{
+		if (!LOCATION.isInScene(plugin.getClient()))
+		{
+			return;
+		}
+
+		for (net.runelite.api.NPC npc : plugin.getNpcsToMark())
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java]: name `NPC` something better.

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollMusicOverlay.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		this.plugin = plugin;
+		this.client = client;
+	}
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		ClueScroll clue = plugin.getClue();
+
+		if (clue == null || !(clue instanceof MusicClue))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollMusicOverlay.java]: `instanceof` is null safe

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import net.runelite.client.ui.overlay.components.TitleComponent;
+import static net.runelite.client.plugins.cluescrolls.ClueScrollOverlay.TITLED_CONTENT_COLOR;
+import static net.runelite.client.plugins.cluescrolls.ClueScrollWorldOverlay.IMAGE_Z_OFFSET;
+
+@Getter
+public class MusicClue extends ClueScroll implements NpcClueScroll
+{
+	private static WorldPoint LOCATION = new WorldPoint(2990, 3384, 0);
+	private static String NPC = "Cecilia";
+	private static Pattern SONG_PATTERN = Pattern.compile("<col=ffffff>([A-Za-z ]+)</col>");
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java]: Songs can contain `'&,` too

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java (hunk 2 above)
Comment (Hydrox6) [reply runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java]: This is so the Clue Plugin will highlight Cecilia

**Comment by Hydrox6** [comment]
Updated the scroll function, since I realised that medium emote clues would want both emotes on screen at the same time.

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import net.runelite.client.ui.overlay.components.TitleComponent;
+
+@Getter
+public class MusicClue extends ClueScroll implements NpcClueScroll
+{
+	private static final WorldPoint LOCATION = new WorldPoint(2990, 3384, 0);
+	private static final String CECILIA = "Cecilia";
+	private static final Pattern SONG_PATTERN = Pattern.compile("<col=ffffff>([A-Za-z !&',.]+)</col>");
+
+	private String song;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/MusicClue.java]: Final. also remove constructor and use @AllArgsConstructor

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollPlugin.java [hunk 7]
```diff
@@ -143,6 +157,8 @@
 	private Integer clueItemId;
 	private boolean worldMapPointsSet = false;
 
+	private final TextComponent textComponent = new TextComponent();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollPlugin.java]: Move this above non-final fields

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollMusicOverlay.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+import net.runelite.api.Client;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.plugins.cluescrolls.clues.ClueScroll;
+import net.runelite.client.plugins.cluescrolls.clues.MusicClue;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+
+public class ClueScrollMusicOverlay extends Overlay
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollMusicOverlay.java]: remove public

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollMusicOverlay.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+{
+	private static final Rectangle PADDING = new Rectangle(2, 1, 0, 1);
+
+	private final ClueScrollPlugin plugin;
+	private final Client client;
+
+	@Inject
+	public ClueScrollMusicOverlay(ClueScrollPlugin plugin, Client client)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollMusicOverlay.java]: private

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollEmoteOverlay.java [hunk 10]
```diff
@@ -40,18 +35,11 @@
 import net.runelite.client.ui.overlay.Overlay;
 import net.runelite.client.ui.overlay.OverlayLayer;
 import net.runelite.client.ui.overlay.OverlayPosition;
-import net.runelite.client.ui.overlay.OverlayUtil;
-import net.runelite.client.ui.overlay.components.TextComponent;
 
 public class ClueScrollEmoteOverlay extends Overlay
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollEmoteOverlay.java]: remove public

**Pull request closed without merging**